### DISCUSSION TITLE
KC26, secondary datasource as user storage provider, AWS JDBC wrapper which does not allow XA

### DISCUSSION THREAD

**Discussion by BerGer23** [keycloak/keycloak#34837] [Feedback]
Hi everybody,

I would like to ask for advice on how to handle the enforcement of XA with Keycloak 26 in a specific scenario.

In the architecture in question, KeycloakX runs in K8s and connects to two Postgres datasources (keycloakdb, userstore).
These two separate databases are provided by the same AWS RDS instance.
`keycloakdb` is the `KC_DB`.
`userstore` is read/written by another service in the cluster which manages it via Liquibase.
From KC's perspective this is the user storage provider and a secondary datasource which is functionally **readonly** (hence the enforcement of XA feels "greedy" in this specific case).

Connections are handled by [AWS Advanced JDBC Wrapper](https://github.com/aws/aws-advanced-jdbc-wrapper), which does not allow using XA (see [this comment](https://github.com/aws/aws-advanced-jdbc-wrapper/issues/511#issuecomment-1624417274)).

The new [checkMultipleDatasourcesUseXA](https://github.com/keycloak/keycloak/blame/637ca2e13848c74444d98604918670217247bf64/quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/KeycloakProcessor.java#L306) prevents this setup from working and blocks the upgrade to v26.

Do you have any advice for this scenario?
Thank you in advance!

**Comment by shawkins** [comment]
> From KC's perspective this is the user storage provider and a secondary datasource which is functionally readonly (hence the enforcement of XA feels "greedy" in this specific case).

Keycloak unfortunately does not know the intent of the additional datasources.

> Do you have any advice for this scenario?

We did not provide a staight-forward workaround to this scenario. The only thing that can be done currently is to wrap one of the Datasources with something that looks like an XADatasource.

You'll probably want to log an enhancement for some mechanism to disable this check in Keycloak.

**Comment by BerGer23** [comment]
Thank you very much for the quick answer @shawkins. 
I will go ahead and open a feature/improvement request.

> The only thing that can be done currently is to wrap one of the Datasources with something that looks like an XADatasource.

Do you know whether there is an example on how to do this in the documentation? 
Currently we do not touch the two datasources programmatically - `keycloakdb` is defined via env variables `KC_DB*`, `userstore` via `quarkus.properties`.

I imagine the workaround would look somewhat similar to [this opentelemetry example](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/jdbc/library#datasource-way), but of course with some XADatasource type as the wrapping class?

**Comment by shawkins** [comment]
> Do you know whether there is an example on how to do this in the documentation?

I don't believe so.

> I imagine the workaround would look somewhat similar to [this opentelemetry example](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/jdbc/library#datasource-way), but of course with some XADatasource type as the wrapping class?

Yes, something like https://docs.oracle.com/cd/E55783_01/Platform.11-2/apidoc/atg/service/jdbc/FakeXADataSource.html but I'm not sure of where there's source available for that or other similar efforts.

**Comment by BerGer23** [comment]
Thanks again, I'll take a look.
Feature request has been logged at https://github.com/keycloak/keycloak/issues/34880.
